src/main/drivers/compass/compass_qmc5883l.c

```diff
@@ -110,7 +111,7 @@ static bool qmc5883lRead(magDev_t *magDev, int16_t *magData)
 
         case STATE_WAIT_STATUS:
```

Comment:
I suggest the following patch to simplify the state machine and tidy things up a little.

```
diff --git a/src/main/drivers/compass/compass_qmc5883l.c b/src/main/drivers/compass/compass_qmc5883l.c
index fbbc05c..6073e08 100644
--- a/src/main/drivers/compass/compass_qmc5883l.c
+++ b/src/main/drivers/compass/compass_qmc5883l.c
@@ -68,6 +68,7 @@
 
 #define QMC5883L_REG_DATA_OUTPUT_X 0x00
 #define QMC5883L_REG_STATUS 0x06
+#define QMC5883L_REG_STATUS_DRDY 0x01
 
 #define QMC5883L_REG_ID 0x0D
 #define QMC5883_ID_VAL 0xFF
@@ -93,39 +94,36 @@ static bool qmc5883lInit(magDev_t *magDev)
 static bool qmc5883lRead(magDev_t *magDev, int16_t *magData)
 {
     static uint8_t buf[6];
-    static uint8_t status;
+    static uint8_t status = 0;
     static enum {
-        STATE_READ_STATUS,
-        STATE_WAIT_STATUS,
-        STATE_WAIT_READ,
-    } state = STATE_READ_STATUS;
+        STATE_WAIT_DRDY,
+        STATE_READ,
+    } state = STATE_WAIT_DRDY;
 
     extDevice_t *dev = &magDev->dev;
 
     switch (state) {
         default:
-        case STATE_READ_STATUS:
-            busReadRegisterBufferStart(dev, QMC5883L_REG_STATUS, &status, sizeof(status));
-            state = STATE_WAIT_STATUS;
-            return false;
-
-        case STATE_WAIT_STATUS:
-            if ((status & 0x01) == 0) {
+        case STATE_WAIT_DRDY:
+            if (status & QMC5883L_REG_STATUS_DRDY) {
+                // New data is available
+                busReadRegisterBufferStart(dev, QMC5883L_REG_DATA_OUTPUT_X, buf, sizeof(buf));
+                state = STATE_READ;
+            } else {
+                // Read status register to check for data ready
                 busReadRegisterBufferStart(dev, QMC5883L_REG_STATUS, &status, sizeof(status));
-                return false;
             }
-
-            busReadRegisterBufferStart(dev, QMC5883L_REG_DATA_OUTPUT_X, buf, sizeof(buf));
-            state = STATE_WAIT_READ;
             return false;
 
-        case STATE_WAIT_READ:
-
+        case STATE_READ:
             magData[X] = (int16_t)(buf[1] << 8 | buf[0]);
             magData[Y] = (int16_t)(buf[3] << 8 | buf[2]);
             magData[Z] = (int16_t)(buf[5] << 8 | buf[4]);
 
-            state = STATE_READ_STATUS;
+            state = STATE_WAIT_DRDY;
+
+            // Indicate that new data is required
+            status = 0;
 
             return true;
     }
```

Reply:
This has been implemented now.